### PR TITLE
Add ConversionFunctionBuilder

### PR DESCRIPTION
**Pull request by matthchr**
- Refactor ConvertTo/From ARM handlers to use it.
    This reduces duplicate code and fixes a small bug in
    the ConvertToARM handler.

### CODE REVIEW

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (matthchr) [inline hack/generator/pkg/astmodel/conversion_function_builder.go:17]: Probably should have comments on these functions/types

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (matthchr) [inline hack/generator/pkg/astmodel/conversion_function_builder.go:125]: It'd be nice to both: "have an order" to things that are processed, but also allow them to be overwritable/movable (like `TypeVisitor`'s are).

I didn't spend a ton of time thinking how to accomplish that for this first pass but possibly something to think about if this is going to be used more broadly.

File: hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go
Comment (matthchr) [inline hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go:54]: I initially wanted to do away with this double-nesting, (we have handlers and then one of the handlers uses the `typeConversionBuilder` which itself has handlers), but actually it makes sense...

The upper handler (`propertyConversionHandlers`) is about determining which properties are involved: given a property on the destination type it determines which property (if any) on the source type will be converted to the destination.
The "inner" handler `typeConversionBuilder` is about determining how to convert between two types: given a source type and a destination type, figure out how to make an assignment. It has no knowledge of broader object structure of "other" properties.

File: hack/generator/pkg/astmodel/armconversion/convert_to_arm_function_builder.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/armconversion/convert_to_arm_function_builder.go:266]: This seems to have a built in assumption that the source and destination properties are the same type, which possibly makes this a simpler application than the conversions I've written for storage types.

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/conversion_function_builder.go:29]: Do you need to call out that it is cloned? It's kind of an implementation detail.

What about making the field `source` and providing a method `Source()` that just happens to return a clone.

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/conversion_function_builder.go:34]: Same idea here
```suggestion
func (params ConversionParameters) Destination() dst.Expr {
```

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/conversion_function_builder.go:91]: For the storage conversions I reduced the number of collisions by using distinct suffixes - `Key` and `Value` for maps, `Item` and `Index` (if I recall correctly) for lists, and so on.

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/conversion_function_builder.go:173]: Suggest using the debug descriptions of the types to provide more information about each

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/conversion_function_builder.go:215]: Use `astbuilder.AddrOf()`

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/conversion_function_builder.go:330]: Maybe `item`?

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/conversion_function_builder.go:398]:
```suggestion
//
// <destination> <assignmentHandler> <source>
//
```

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/conversion_function_builder.go:471]: Use `astbuilder.CallQualifiedFunc()`

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (theunrepentantgeek) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:330]: Or if this is the list, how about taking the singular form of the property name and suffixing `List` ?

**Review by theunrepentantgeek — CHANGES_REQUESTED** [review]

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (babbageclunk) [inline hack/generator/pkg/astmodel/conversion_function_builder.go:202]: I think this would be clearer if you just construct a new `ConversionParameters` and explicitly copy some fields from `params` and set some with new values. Otherwise you have to refer back to the struct def to work out which fields are the same and which are different.

**Review by babbageclunk — APPROVED** [review]
Looks good - although I agree with @theunrepentantgeek's suggestions too.

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:109]: I think you misread this as 
`result.ConversionContext = append(result.ConversionContext, result.ConversionContext...)`
and not
`result.ConversionContext = append(result.ConversionContext, params.ConversionContext...)`

It's not duplicating the slice it's appending the `params` `ConversionContext` to `result`'s

File: hack/generator/pkg/astmodel/armconversion/convert_to_arm_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/armconversion/convert_to_arm_function_builder.go:52]: Because of the `definedErrVar` which I'm currently using to stick a `var err` at the top of functions that need it.
We need this because the following isn't valid code:
```
err := populateFromARM(foo)
err := populateFromARM(bar)
```

The current way I have things is kinda icky though because it means that these top level handlers need to understand a bit about what the lower level handler is going to do in order to determine if it needs to start with that `var err` or not. I think I see a way to fix this and make the `from` handler the same as the `to` handler... will investigate

File: hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go:54]: Done

File: hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/armconversion/convert_from_arm_function_builder.go:219]: Done (across all occurrences here and in the `_to_arm` builder as well).

File: hack/generator/pkg/astmodel/armconversion/convert_to_arm_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/armconversion/convert_to_arm_function_builder.go:266]: Do you specifically mean the optional property handling here? Yes, in the convert to/from ARM case optionality is preserved between the two types always (no loss of information possible).

Definitely agree that these are some hurdles we would have to cross before merging the infrastructure for these two types of handlers (storage vs ARM). Part of merging them would also probably be more fine-grained control so that each usage (storage vs arm) could easily override the bits that made sense for it.

For now though I just maintained this behavior as that's what I had before and it wasn't clear that there was a generic solution that would work for both - which is to say I suspect the storage cases where only one side is optional we'd need handlers for those specific cases (in addition to the default "both" case)?

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:29]: Currently this is in the astmodel package and the ARM stuff is in a subpackage, so I can't make these fields private (not exported?) as that would prevent the usage from constructing them.

I could change the methods to `GetSource` + `GetDestination`?

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:34]: See comment above.

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:173]: We don't have easy access to the full `types` collection here... we do have `builder.CodeGenerationContext.GetTypesInCurrentPackage()` but that's only the types in _this_ package so won't work for types in other packages.

I added `CodeGenerationContext.GetAllReachableTypes()` and used that instead.

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:202]: Done

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:215]: Done - fixed other usages of UnaryExpr too

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:398]: Fixed

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:471]: `CallQualifiedFunc` expects that I have a `string` for `qualifier` but here I have an `Expr`. Given that seems to be an "abnormal" case (we have a good number of other usages of `CallQualifiedFunc`, and the fact that we have helpers for the other bits here I'm not sure it makes sense adding a helper to `astbuilder` to cater to this scenario.

The code I have right now is:
```
newSource := astbuilder.Dereference(
		&dst.CallExpr{
			Fun: astbuilder.Selector(params.ClonedSource(), "DeepCopy"),
			Args: []dst.Expr{},
		})
```

which isn't too bad?

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:91]: I'd like to make some improvements to the naming, but I think I should leave that part for a follow up PR, as any changes to how we generate names in the code is going to explode the PR side in the generated files - thoughts?

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:330]: Same comment as above about holding off on generated name changes until a second PR to ease review burden.

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:109]: Nevermind ignore the above - you're right my own code confused me too. Changed.

File: hack/generator/pkg/astmodel/armconversion/convert_to_arm_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/armconversion/convert_to_arm_function_builder.go:52]: I have a fix for this - it's with a bunch of other renaming stuff though so will include it in a second PR I think

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (theunrepentantgeek) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:30]: What if you introduced a constructor/factory method so that direct construction wasn't needed?

If that's not feasible, then `GetSource()` etc would be fine.

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (theunrepentantgeek) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:92]: Followup PR is fine.

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (theunrepentantgeek) [reply hack/generator/pkg/astmodel/conversion_function_builder.go]: Yeah, that's better than it was. 👍

**Review by theunrepentantgeek — APPROVED** [review]

File: hack/generator/pkg/astmodel/conversion_function_builder.go
Comment (matthchr) [reply hack/generator/pkg/astmodel/conversion_function_builder.go:30]: The issue with constructor/factory method is that this type has 8 properties, some of which are regularly `nil`. An 8 parameter function is significantly less understandable/readable than an 8 parameter struct instantiation I think.

One approach that could get these names nicer would be to duplicate the `ConversionParameters` structure and turn it into a `parameters` struct that's only used to supply arguments to the factory function. Then make everything not exported on `ConversionParameters` so that we could use `Source` and `Destination`. That's quite a bit of extra code though and feels like it might be a bit overengineered.

Will just change to `GetX` for now I think

**Pull request merged**
